[Green CI] Fix test test_storage_s3_queue/test.py::test_max_set_age

pamarcos opened the pull request:
Fix `test_storage_s3_queue/test.py::test_max_set_age`

- Use unique filenames to prevent following tests failing.
- Use a busy-loop instead of a sleep to find the condition is met ASAP.
- Increase from 10s to 20s the max_age. Fixes several issues for time-sensitive checks. For instance, in case the CPU is more overloaded than usual (for instance, because it's executing several heavy parallel tests), the time elapsed between some close test steps might get even surpass max_age when it's not expected to do so. This makes the S3Queue to do an extra insertion, invalidating lot of the assumptions of the test.
- Remove redundant checks.

Apart from fixing the test itself:
- Update integration test documentation, adding more information about parallel tests
- Use a different instance name when using `pytest-xdist` to allow parallel test execution 

I've run hundreds of times the test in parallel with the following command with no flakiness found:

```bash
tests/integration/runner test_storage_s3_queue/test.py::test_max_set_age -- --count 10000 -n 16
```

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

Closes [#13110](https://github.com/ClickHouse/clickhouse-private/issues/13110)

pamarcos committed:
Use unique names for tables and files

pamarcos committed:
Refactor test to improve it

- Create wait_for_condition that checks greedily for a period of time
- Remove redundant checks
- Allow other tests running in parallel to have `ObjectStorageQueueFailedFiles` errors

alexey-milovidov commented on tests/integration/test_storage_s3_queue/test.py:
```diff
@@ -847,107 +848,85 @@ def test_max_set_age(started_cluster):
         "\n".join((",".join(map(str, row)) for row in values)) + "\n"
     ).encode()
-    put_s3_file_content(started_cluster, f"{files_path}/fff.csv", values_csv)
+    file_with_error = f"fff_{uuid4().hex}.csv"
```
Don't we use unique containers for every test and a unique Minio instance for every container?

alexey-milovidov replied:
If not, you can also fix this test: https://github.com/ClickHouse/ClickHouse/pull/66802/files

pamarcos replied:
That's right AFAIU, but I couldn't use `--count 1000` to run successfully a bunch of tests sequentially without this modification. As expected, between each test execution ClickHouse's system tables are kept, so the `count()` check against `system.s3queue_log` fails if we always use the same filename.

I'll try to investigate how I can run in parallel N docker containers running the same test, because I tried but failed because some files are apparently shared by pytest.

kssenii commented:
> Use unique names for tables and files to prevent parallel tests affecting. This is what actually fixes the test

AFAIK tests in a single test.py can be parallelized, but in this case they will be run with different clickhouse servers. And within each such batch of tests which is run on a single clickhouse server, they are NOT run in parallel. So it does not matter what the names of the tables are as long as they are dropped after the test or before each further test, which is already done.
Frankly speaking I think if this test fails, then it is not a bug in the test, but a bug in the code.

qoega commented on tests/integration/test_storage_s3_queue/test.py:
```diff
@@ -822,7 +823,7 @@ def get_count(node, table_name):
+    table_name = f"max_set_age_{uuid4().hex}"
     dst_table_name = f"{table_name}_dst"
     keeper_path = f"/clickhouse/test_{table_name}"
     files_path = f"{table_name}_data"
```
IMHO it is much easier to just have separate path for each test run and not change filenames when we have many of them

kssenii commented on tests/integration/test_storage_s3_queue/test.py:
```diff
@@ -822,7 +823,7 @@ def get_count(node, table_name):
 def test_max_set_age(started_cluster):
     node = started_cluster.instances["instance"]
-    table_name = f"max_set_age"
+    table_name = f"max_set_age_{uuid4().hex}"
```
there is also no other table created with name "max_set_age" as it was before so it was already unique, isn't it?

qoega replied:
With `--count 10` we run test with same clickhouse and minio. I.e. we need to remove tables we created during a test to allow next run to be successful. 
To simplify work with minio we do not remove anything, but use different paths in each test run.

kssenii replied:
> With --count 10

but we do not use it in CI AFAIK, unless integration tests flaky check may be, but it is run only if the test is changed in the PR

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #67035 from pamarcos/ClickHouse:fix-flaky-test_storage_s3_queue-test_max_set_age -->
*This is an automated comment for commit 91b7001df6e827f801bd792e7bd9d96cdd947946 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67035/91b7001df6e827f801bd792e7bd9d96cdd947946/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67004/2dd97b481de4e2cf34d4b70feb246e179a2edae0/builds/report.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/65521/9d7bb2a430b142e0761cc30efcf78ea1bbeb6871/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/911104dcb2865cd38a92aa45f87e9799207199f9/fast_test.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/7280118845d73613fd8d05ea629146213f7a4ee7/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Format with black

pamarcos committed:
Increase sleep time make sure there is a new failure

The previous sleep was already adding +5s to make sure
the TTL was properly applied, so we'd rather use the
same value here instead of just 1s.

pamarcos committed:
Set a different instance dir when using pytest-xdist

This allows executing in integration tests in parallel
without directory clashes.

pamarcos committed:
Add parallel integration test execution to doc

pamarcos committed:
Clarify documentation

pamarcos committed:
Fix test flakiness

- Increase from 10s to 20s the max_age. Removes flakiness due to
  the files in minio still present in following execution, even
  if the cleanup stage of the test is clearly removing them.
- Use a busy-loop to check conditions directly, removing the
  unnecessary sleep.

pamarcos committed:
Remove logging traces to help debugging

pamarcos committed:
Merge branch 'master' into fix-flaky-test_storage_s3_queue-test_max_set_age

pamarcos commented:
I've left my AWS dev machine all night running the test and I got 8300 successes and no failures.
I checked out at https://github.com/ClickHouse/ClickHouse/pull/67035/commits/5b9427dc566b1e76a82be9f650b9dbc386567fc4 to have some custom traces to help debugging in case something went south.

This is the script that I used, for reference. Running the test in 10 parallel processes in batches of 100:

```bash
!/bin/bash

set -uo pipefail

if [[ $(docker ps | grep -v CREATED) ]]; then
    docker ps | grep -v CREATED | awk '{print $1}' | xargs docker kill
fi

export CLICKHOUSE_TESTS_CLIENT_BIN_PATH=/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse
export CLICKHOUSE_TESTS_SERVER_BIN_PATH=/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse
export CLICKHOUSE_TESTS_BASE_CONFIG_DIR=/home/ubuntu/ClickHouse/ClickHouse/programs/server

find tests/integration -iname "pytest-gw*.log" 2> /dev/null | xargs rm || true

for _ in {1..1000}; do
    echo "Starting batch of 1000 tests..."
    tests/integration/runner test_storage_s3_queue/test.py::test_max_set_age -- --count 100 -n 10 2>&1 | tee -a test_max_set_age.log
    echo "Log results..."
    grep -inR "PMO" tests/integration/pytest*.log | tee -a test_max_set_age.log
done
```

And here's the log file: 
[test_max_set_age.zip](https://github.com/user-attachments/files/16424730/test_max_set_age.zip)

kssenii commented on tests/integration/test_storage_s3_queue/test.py:
```diff
@@ -847,107 +848,87 @@ def test_max_set_age(started_cluster):
-            "SELECT value FROM system.events WHERE name = 'ObjectStorageQueueFailedFiles'"
-        )
-    )
+    wait_for_condition(lambda: failed_count + 2 <= get_object_storage_failures())
```
The point here, as I remember, was to check that in `max_set_age` time we increase failed_count only by 1, but now it does not check this (as it accepts more than 1).

kssenii replied:
ah, you explained it in PR description...

> Other tests running in parallel may have any object storage failure, making the equality with exactly +1 or +2 not to be true anymore.

but I thought it is impossible that other tests can be run in parallel with the same clickhouse server. AFAIK tests run in parallel use different clickhouse servers.

pamarcos replied:
Yep, that's an unfortunate side-effect. In my case it was convenient to change it to be able to run several of this test in parallel through `pytest-xsdist` without affecting each other. But then I thought that on the CI you cannot ensure no other test contributes to this error unless we completely isolate the ClickHouse DB for this one. Otherwise, how can we ensure that `system.events` is never going to contain any `ObjectStorageQueueFailedFiles` created by any other test?

AFAIU how the CI works, it runs 5 of them in parallel. We know for sure all tests in `test_storage_s3_queue` never run in parallel. But what about the others? There are other 4 tests running on the same ClickHouse DB that could potentially affect `ObjectStorageQueueFailedFiles`. The alternative is to be super organized and ensure anything that may affect it lives within the same test so that they're never parallelized.

kssenii replied:
> There are other 4 tests running on the same ClickHouse DB that could potentially affect ObjectStorageQueueFailedFiles

But AFAIK no other tests suit apart from `test_storage_s3_queue` uses S3Queue (AzureQueue tests are in the same `test_storage_s3_queue`), so as you wrote "We know for sure all tests in test_storage_s3_queue never run in parallel." leads to a conclusion that `ObjectStorageQueueFailedFiles` cannot be changed anywhere apart from this test itself when it is executing.

pamarcos replied:
[Rolled back ](https://github.com/ClickHouse/ClickHouse/pull/67035/commits/15e0033016eb0e23a7e6f512d5096e50863e3187) the strict check because I think you're right.

I swear at some point I got assertion errors on that check when running in parallel, but maybe it was not the exactly same setup I have now which is closer to the CI. In any case, sorry for the noise. I'm running in my dev machine a batch of tests to make sure this still works ok, just in case.

pamarcos replied:
The good news is that I collected a list of assertions that failed for the test [here](https://github.com/ClickHouse/clickhouse-private/issues/13110#issuecomment-2250520981) and this assert is not among them. It's also not the one in the description of the issue.

The TL;DR is that increasing from 10s to 20s the max_age is what fixes most of the asserts :roll_eyes:

pamarcos replied:
I've run it with 100% successful rate after +1000 test executions :rocket:

pamarcos committed:
Bring back the strict check

Also update the doc.

tavplubix commented:
@pamarcos, looks like it breaks `test_zookeeper_config_load_balancing` somehow

antaljanosbenjamin commented:
I find it strange that the CI Check is green, however the integration tests were still running as far as I see.

pamarcos commented:
> @pamarcos, looks like it breaks `test_zookeeper_config_load_balancing` somehow

It's because of the change in [cluster.py](https://github.com/ClickHouse/ClickHouse/pull/67035/files#diff-ee7374aefb404c931cf5a282f81454b08a48ad4bf47b74739dd2f0f709cae9f5R438-R442). Let me create a new PR to remove those lines and I'll investigate in a separate PR how to make that particular test work while having them.

I also find it very strange that the CI for this PR didn't catch it.
